docs: add MTEB evaluation guide and update usage.rst

sahibpreetsingh12 opened the pull request:
This PR resolves [#3332](https://github.com/UKPLab/sentence-transformers/issues/3332).

### Summary

Adds a new documentation page for evaluating SentenceTransformer models using the [Massive Text Embedding Benchmark (MTEB)](https://github.com/embeddings-benchmark/mteb), along with relevant task examples and best practices.

### Changes

- `mteb_evaluation.md` in `docs/sentence_transformer/usage/`:
  - Installation steps
  - Minimal working example
  - Task-type breakdown (STS, Classification, Retrieval, etc.)
  - Notes on output handling
  - Warnings about *not* using MTEB during training
  - Leaderboard + export instructions

-  Linked from `usage.rst` to include in sidebar navigation

### Notes

Following the guidance in the discussion, MTEB is documented as a **post-training evaluation tool**, not integrated as an evaluator to avoid benchmark overfitting.

Let me know if you'd like any section adjusted. Thank you!

sahibpreetsingh12 commented:
@tomaarsen Please share your feedback and if anything else i should chnage

Samoed commented on docs/sentence_transformer/usage/mteb_evaluation.md:
```diff
@@ -0,0 +1,135 @@
+from mteb import MTEB
+
+model = SentenceTransformer("all-MiniLM-L6-v2")
+evaluation = MTEB(tasks=["STSBenchmark"])
```
Recommended way
```suggestion
tasks = mteb.get_tasks(tasks=["STSBenchmark"])
evaluation = MTEB(tasks=tasks)
```

Samoed commented on docs/sentence_transformer/usage/mteb_evaluation.md:
```diff
@@ -0,0 +1,135 @@
+
+The [Massive Text Embedding Benchmark (MTEB)](https://github.com/embeddings-benchmark/mteb) is a comprehensive benchmark suite for evaluating embedding models across diverse NLP tasks like classification, retrieval, clustering, reranking, and semantic similarity.
+
+This guide walks you through using MTEB **with SentenceTransformer models for post-training evaluation**. This is *not* designed for use during training loops.
```
Can you add something like?
```suggestion
This guide walks you through using MTEB **with SentenceTransformer models for post-training evaluation**. This is *not* designed for use during training loops. To fully integrate your model to `MTEB` you can follow this [guide](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_model.md)
```

Samoed commented on docs/sentence_transformer/usage/mteb_evaluation.md:
````diff
@@ -0,0 +1,135 @@
+You can evaluate specific tasks or full categories:
+
+```python
+MTEB(tasks=["SICK-R", "AmazonCounterfactualClassification"])
+MTEB(task_types=["Classification", "Retrieval"])
````
```suggestion
mteb.get_tasks(tasks=["SICK-R", "AmazonCounterfactualClassification"])
mteb.get_tasks(task_types=["Classification", "Retrieval"])
```

Samoed reviewed (commented):
@KennethEnevoldsen can you look too?

sahibpreetsingh12 commented:
@Samoed  and @tomaarsen 
If anything else is required form my side please do share.

Samoed commented on docs/sentence_transformer/usage/mteb_evaluation.md:
````diff
@@ -0,0 +1,136 @@
+Install MTEB and its dependencies:
+
+```bash
+pip install mteb
````
We're now working on `v2` release that would be breaking
```suggestion
pip install "mteb<2"
```

KennethEnevoldsen replied:
I don't see anything that wouldn't work in v2 as well?

Samoed replied:
Everything would work, but `mteb.MTEB` will be deprecated

KennethEnevoldsen replied:
It will simply display a deprecation warning (with an upgrade suggestion), which I think is perfectly fine.

sahibpreetsingh12 replied:
so @KennethEnevoldsen  and @Samoed  guys what's your final suggestion shall i change dependencies or no?

KennethEnevoldsen replied:
Keep it as is

Samoed commented on docs/sentence_transformer/usage/mteb_evaluation.md:
```diff
@@ -0,0 +1,136 @@
+from sentence_transformers import SentenceTransformer
+from mteb import MTEB
+
+model = SentenceTransformer("all-MiniLM-L6-v2")
```
I'm not sure what the best approach is here, because this documentation is for `SentenceTransformer`, not `MTEB`. However, this method might be better since some models could have custom instructions or implementations coming from `MTEB`.

```suggestion
model = mteb.get_model("sentence-transformers/all-MiniLM-L6-v2")
```

sahibpreetsingh12 replied:
ok so i have left it 
at 
`model = SentenceTransformer("all-MiniLM-L6-v2")`

KennethEnevoldsen commented on docs/sentence_transformer/usage/mteb_evaluation.md:
```diff
@@ -0,0 +1,136 @@
+
+### 1. **Semantic Textual Similarity (STS)**
+
+* **STSBenchmark**: Predict similarity score (0-5) for English sentence pairs.
```
It is not clear that these are just examples - it is not a comprehensive list

sahibpreetsingh12 replied:
I have now added a clear Note to avoid any further confusion

KennethEnevoldsen commented on docs/sentence_transformer/usage/mteb_evaluation.md:
````diff
@@ -0,0 +1,136 @@
+
+```python
+mteb.get_tasks(tasks=["SICK-R", "AmazonCounterfactualClassification"])
+mteb.get_tasks(task_types=["Classification", "Retrieval"])
````
This would often get you more than a 100 tasks. I would never recommend it. It is however a good way to find if there are tasks that matche your use case - e.g. if we want to check if there is any medical retrieval tasks for english: 

```
# this needs to be validated:
tasks = mteb.get_tasks(task_types=["Retrieval"], domains = ["Medical"], languages=["eng"])
```

KennethEnevoldsen commented on docs/sentence_transformer/usage/mteb_evaluation.md:
````diff
@@ -0,0 +1,136 @@
+```python
+results = evaluation.run(model, output_folder=None)
+
+from mteb import MTEBResults
+summary = MTEBResults(results).main_scores()
+print(summary)
````
this kinda hides what happens - I would just extract the main score pr task and average them

KennethEnevoldsen commented on docs/sentence_transformer/usage/mteb_evaluation.md:
````diff
@@ -0,0 +1,136 @@
+To avoid writing results to disk:
+
+```python
+evaluation.run(model, output_folder=None)
````
I would always save the results (`v2` will automatically save it to a cache directory, which you of course can specify)

sahibpreetsingh12 replied:
Yes @KennethEnevoldsen  did it changes done now

KennethEnevoldsen commented on docs/sentence_transformer/usage/mteb_evaluation.md:
```diff
@@ -0,0 +1,136 @@
+
+* Using it during training risks **overfitting** to public benchmarks.
+* It writes to disk and caches aggressively.
+* Official guidance recommends using SentenceTransformer's built-in evaluators like:
```
If you want to evaluate during training?

sahibpreetsingh12 replied:
@KennethEnevoldsen ok so here the thought was to just show we do evaluation at testing
Although i also thought about it but then idea was to keep things simple and if this works
I would open another issue and then get it done.

KennethEnevoldsen replied:
Hmm, I am not quite sure what you mean, but from my reading, it sounds like you do not recommend MTEB for evaluation (even after training), which I don't think is the intention?

sahibpreetsingh12 replied:
Thanks for the clarification, @KennethEnevoldsen !

You're absolutely right — my intention was not to discourage the use of MTEB for evaluation after training. I’ve now updated the wording to clarify that MTEB is **recommended for post-training evaluation**, but **not ideal during training loops** due to the risk of overfitting and aggressive caching.

Let me know if the revised phrasing works better — happy to tweak further! 🙌

KennethEnevoldsen commented on docs/sentence_transformer/usage/mteb_evaluation.md:
```diff
@@ -0,0 +1,136 @@
+**Important**: MTEB is for *post-training* benchmarking only.
+
+* Using it during training risks **overfitting** to public benchmarks.
+* It writes to disk and caches aggressively.
```
Is this not ideal?

sahibpreetsingh12 replied:
yes @KennethEnevoldsen  from my understanding and based on discussion we had at the start
This looks ideal for me.

KennethEnevoldsen commented on docs/sentence_transformer/usage/mteb_evaluation.md:
````diff
@@ -0,0 +1,136 @@
+
+```python
+from mteb import MTEBResults
+df = MTEBResults(results).to_markdown()
+print(df)
````
Maybe it is better to use the "to_dataframe" method instead?

sahibpreetsingh12 replied:
ok @KennethEnevoldsen  I have moved it to "to_dataframe" method

KennethEnevoldsen commented on docs/sentence_transformer/usage/mteb_evaluation.md:
````diff
@@ -0,0 +1,136 @@
+print(df)
+```
+
+Follow submission instructions in the [MTEB repo](https://github.com/embeddings-benchmark/mteb).
````
```suggestion
To add your results to the MTEB Leaderboard, follow the submission instructions in the [MTEB repository](https://github.com/embeddings-benchmark/mteb).
```

sahibpreetsingh12 commented on docs/sentence_transformer/usage/mteb_evaluation.md:
````diff
@@ -0,0 +1,136 @@
+evaluation.run(model, output_folder="results/")
+```
+
+This evaluates your model on the **STS Benchmark**, a Semantic Textual Similarity dataset with human-annotated sentence pairs. Output is saved in `results/`.
````
@KennethEnevoldsen any particluar task which will be better for this documentation
Just let me know will push the changes

KennethEnevoldsen replied:
STS17 or STS22.v2 are both pretty good. Both are multilingual by default, well-known tasks in the field (SemEval tasks), and are relatively small, so the evaluation won't take too long.

sahibpreetsingh12 replied:
I have taken STS22.v2 @KennethEnevoldsen

sahibpreetsingh12 commented:
@KennethEnevoldsen and @Samoed 

All review suggestions have been incorporated:
- Replaced STSBenchmark with STS22.v2
- Clarified task examples with disclaimer and link to full list
- Added filtering by task_type, domain, language
- Used `.to_dataframe()` for result printing
- Explicitly looped over `main_score` values

Let me know if anything else is required — happy to revise 🙌

tomaarsen commented:
Hello @sahibpreetsingh12,

Thanks for the PR. I've had a detailed look at it now, and I'm afraid I see a lot of issues/quirks. For example:
* Very little of code the runs
* The code snippets don't match the `mteb` suggested format of `import mteb`, `mteb.get_tasks`, but instead use e.g. `from mteb import get_tasks`
* There's text in a code block in the Quick Start
* The section titles are unusually long compared to related documentation in Sentence Transformers, they also use emojis unlike other documentation
* There's links to Sentence Transformers documentation, even though this *is* Sentence Transformers documentation
* The list of tasks/examples is a bit arbitrary
* There's a second file under `.ipynb_checkpoints` that likely should not have been included

Overall, the documentation page reads very "AI generated", which I'd like to avoid, if possible. I've overhauled it now, with working code, etc. The overall message from your version still remains.

Thank you @Samoed  and @KennethEnevoldsen for taking the time to review the PR here thus far.

Feel free to let me know what you think of the new version, I think we're pretty close to ready now.

- Tom Aarsen

sahibpreetsingh12 commented:
> Hello @sahibpreetsingh12,
> 
> Thanks for the PR. I've had a detailed look at it now, and I'm afraid I see a lot of issues/quirks. For example:
> 
> * Very little of code the runs
> * The code snippets don't match the `mteb` suggested format of `import mteb`, `mteb.get_tasks`, but instead use e.g. `from mteb import get_tasks`
> * There's text in a code block in the Quick Start
> * The section titles are unusually long compared to related documentation in Sentence Transformers, they also use emojis unlike other documentation
> * There's links to Sentence Transformers documentation, even though this _is_ Sentence Transformers documentation
> * The list of tasks/examples is a bit arbitrary
> * There's a second file under `.ipynb_checkpoints` that likely should not have been included
> 
> Overall, the documentation page reads very "AI generated", which I'd like to avoid, if possible. I've overhauled it now, with working code, etc. The overall message from your version still remains. I hope that's okay.
> 
> Thank you @Samoed and @KennethEnevoldsen for taking the time to review the PR here thus far.
> 
> Feel free to let me know what you think of the new version, I think we're pretty close to ready now.
> 
> * Tom Aarsen

Since these are my early days doing a documentation and even any open source.
Yes I took some help in checking how we can make a documentation but the help was setting a base template but yes I would take it and would definitely improve from here.
And these suggestions help me improve

Thanks @tomaarsen @KennethEnevoldsen  and @Samoed
